rfcs/0000-type-stability.md

```diff
@@ -0,0 +1,501 @@
+[3.5-playground]: https://www.typescriptlang.org/play/?ts=3.5.1&ssl=1&ssc=1&pln=14&pc=27#code/KYDwDg9gTgLgBAE2AMwIYFcA28DGnUDOBcAYqgHY4CeAggghOXAN4CwAUHHGpVXALxwYUdMADcHAL4cOydJRgBLRnEwQIAa3RgAPABUAfAAoCEALbAAaqkyiAXHD0BKFhy6LkcE+as3RAwXIsTBc2Ti44KGAYdCgmcmAAdzgABShzRQJgIyMo00wAN2AXfgNXcIjI4Hyi7wtrW2KJCrhJJ2auaXComLjU9LNM4AA6PIhC7ITksl46BnIjJ3apGXYAejXHAAkASQBlOEsaABkAVQBROABhLZoAOQBxc4OAIXO9AHVz87u4AGZhgAWOD3AAi-2GAFYODhGAR4Mh1AJVOotGAjAAiREQDHtIA
+
+If an addon author wanted to migitate this change, she would follow this procedure:
+
+1.  Backwards-compatibly *fix* the types by explicitly setting the return type on `lookup`:
+
+    ```diff
+    - function lookup<T>(someValue: T) {
+    + function lookup<T>(someValue: T): Promise<{}> {
+    ```
+
+2.  Create a new directory, named something like `ts3.5`.
+
+3.  Generate the type definition files for the addon by running `ember ts:precompile`.
+
+4.  Manually move the generated type definition files into `ts3.5`.
+
+5.  In the `ts3.5` directory, either *remove* or *change* the explicit return type, so that the default from TypeScript 3.5 is restored:
+
+    ```diff
+    - function lookup<T>(someValue: T): Promise<{}> {
+    + function lookup<T>(someValue: T): Promise<unknown> {
+    ```
+
+6.  Commit the `ts3.5` directory, since it now needs to be maintained manually until a breaking change of the library is released which opts into the new behavior.
+
+7.  Cut a release which includes the new fixes. With that release:
+
+    -   Inform users about the incoming breaking change.
+
+    -   Tell users to add `import 'fancy-addon/ts3.5';` to the top of their app's `types/my-app.d.ts` or `types/my-addon.d.ts` file (which is generated by `ember-cli-typescript`’s post-install blueprint).
+
+8.  At a later point, cut a breaking change which opts into the TypeScript 3.5 behavior.
+
+    -   Remove the `ts3.5` directory from the repository.
+
+    -   Note in the release notes that users who did not previously opt into the changes will need to do so now.
+
+    -   Note in the release notes that users who *did* previously opt into the changes should remove the `import 'fancy-addon/ts3.5';` import from `types/my-app.d.ts` or `types/my-addon.d.ts`.
+
+### Policy for supported TypeScript versions
+
+TypeScript and Ember both have regular cadences for release:
+
+- Ember releases every six weeks, with new Long-Term Support (LTS) releases every 24 weeks (that is, every 4 releases).
+- TypeScript releases roughly quarterly. There are no LTS releases for TypeScript.
+
+These map reasonably well to each other, and this correspondence provides a helpful basis for recommending a version support policy.
+
+Addons should generally support the TypeScript versions current during the lifetime of the Ember LTS versions they support. As noted above, dropping a supported TypeScript version is a breaking change, and it is helpful to minimize the number of major versions of an addon in use.
```

Comment:
As I've been mulling on this, I'm debating a bit: if we use *current* LTS instead of *all active LTS*, we basically end up supporting the last *two* TypeScript versions instead of the last *four*—and since TS does *not* do any support for previous versions, we're  not getting a *ton* of benefit out of the support for the extra versions, other than decreasing the *recommended* supported range.

I had this top of mind as I was thinking about machty/ember-concurrency#357 and ember-modifier/ember-modifier#41; the difference is just including 3.6 or not (b/c we need at least two versions supported, so it needs to be 3.7 and 3.9 since 3.8 was broken for us)—but that's actually a pretty nice thing to be able to step past.

Does it actually make sense to support TS versions going back to the *security fixes* LTS version as well as the *active support* LTS? I'm leaning toward *no*, with a qualification like "always at least two supported versions recommended at any given time."

Thoughts, @jamescdavis @dfreeman @rwjblue @pzuraq?

Reply:
I'm inclined to agree—I think aligning with the current active-support LTS is probably reasonable, especially since you're locking in to that as your baseline until the next time you release a major version.

Reply:
Tweaked in af3f9b6 to suggest currently-active Ember LTS will be the normal recommendation, while keeping other verbiage the same.